### ISSUE TITLE
Scorecard showing old results

### ISSUE THREAD

**Issue by sammachin** [node-red/flow-library#89]
node-red-contrib-midi was updated to 1.1.2 on 13/05/2022 with fixes for the license and versions,

However the scorecard ran but still showed the incorrect results
![Screen Shot 2022-05-16 at 3 09 44 PM](https://user-images.githubusercontent.com/187645/168620587-7c090907-41a9-4e59-b5aa-58b903de6ecb.png)

Validation locally with node-red-dev shows that it should have passes for license and versions.

**Comment by pbrunier** [comment]
I have the same issue. Yesterday I published a new version with correct license and today the scorecard still reflects the old version.  https://flows.nodered.org/node/node-red-contrib-dyson-purelink

Running node-red-dev locally shows it is ok:
   node-red-contrib-dyson-purelink@0.6.1
✅ Package is MIT licensed

**Comment by marcus-j-davies** [comment]
Hello,

I apologize if this seems pushy, its not intended to be, only many many Node devs including my myself, have made improvements to our packages, as we get scolded by the scorecard.

Only to find out the scorecard is not updating and our packages still being scored down (aggressively) despite meeting certain standards. 

Many many people on the forums also raise this same issue, and I have also added comments, onto various discussions that are on the forums, without much feedback on the issue.  

This issue is now more than a year old, and I cant see much in the way of any focus into getting it sorted.

I was looking through the codebase of the flow library, to try and get it fixed myself (as I could imagine your all busy), but seems the codebase is more intricate then I can understand.

The scorecard does, I feel, make quite a difference into the level of quality seen by users.

Once again, sorry if this seems flippant - it really isn't meant to , but maybe an update on the issue will be welcome, given the impact a scorecard can have.

**Comment by knolleary** [comment]
You are quite right @marcus-j-davies - we need to get to the bottom of this. I'll revisit it this week.

**Comment by marcus-j-davies** [comment]
Thanks @knolleary,

I'm all for prioritisation, but this one has been going on for sometime, and seeing it crop up more often lately, plus I released a 2nd Node, only to receive 3rd degree burns by the scorecard, and subsequently corrected of course, but not earning the "well done" badge 😅

**Comment by knolleary** [comment]
I've just pushed some extra debug into production to see if I can make any sense of what is going on.

What is truly confusing is that the way this works is:

1. flow library downloads the tar file of the module, unpacks it, extracts all its information and updates database.
2. it then points the scorecard tool at the *same directory* that contains the just-downloaded version - but for some reason the scorecard 'sees' a different version.

Hoping the extra debug I've added sheds some light.

**Comment by knolleary** [comment]
12 hours later, we've had a number of nodes refresh in the library and none have exhibited the mismatch behaviour... so we continue to wait.

**Comment by sammachin** [comment]
@knolleary I just published some updates to my package `@sammachin/node-red-matter-bridge` the first update to 0.0.3 refreshed the scorecard correctly but then a subsequent update to fix the package.json for the node-red version in 0.0.4 and 0.0.5 doesn't seem to have been picked up.

**Comment by knolleary** [comment]
Thanks @sammachin 

From the debug I added last night I can see the following:

```
Scorecard package.json: @sammachin/node-red-matter-bridge 0.0.5
---Validating Package---
@sammachin/node-red-matter-bridge@0.0.3
```
The first line is a new line I added - it loads the package.json file on disk and confirms the version number inside it - 0.0.5.
The last line is from the scorecard tool - reporting it has evaluated 0.0.3.

I have logged into the VM hosting the flow library and checked the files on disk - sure enough only 0.0.5 is present. And when I manually run the scorecard with the same args as the flow library used, I get the 0.0.5 scorecard.

I'm going to have to patch the scorecard tool with some extra debug to figure this out.

**Comment by sammachin** [comment]
thats very odd, if I remember correctly it downloads the package to a tmp folder with a unique name too? (some random chars added to the end) 
I wonder if the scorecard tool isn't closing things properly after the first run?

**Comment by marcus-j-davies** [comment]
Another avenue  that maybe of interest.

The checks make a lot of `require` calls.
On a historical project of mine I needed a a way to invalidate the nodejs cache via `delete require.cache[require.resolve(...)]`


> to a tmp folder with a unique name too?

I don't think it does, but then I'm  not that familiar with the codebase
https://github.com/node-red/node-red-dev-cli/blob/2595fbb2bfc4b96f3de0963ea660831c8f79122f/src/libs/npmget.js#L7

**Comment by knolleary** [comment]
@marcus-j-davies 🎉  that'll explain it.

**Comment by marcus-j-davies** [comment]
Now now Nick! - could be a red herring 😅

**Comment by knolleary** [comment]
Okay - 0.1.6 of node-red-dev published with the fix.

Flow Library updated to use that... deploy in progress. Now we wait and see. If this theory is right (and I'm 99.9% sure it is) then we have to wait for a module to do its *second* refresh after this restart to confirm all is well.

**Comment by sammachin** [comment]
> I don't think it does, but then I'm not that familiar with the codebase

@marcus-j-davies The way the flow library works is different to locally is that it has already downloaded the package from npm so the scorecard tool is pointed at the local folder rather than downloading from NPM itself.
The scorecard tool is run as part of the flow library process then so I can see that the require cache would be an issue here that we don't pickup when running it from a shell as a new process.

**Comment by marcus-j-davies** [comment]
@knolleary,

I am afraid this hasn't worked,

I have just pushed 2 updates the first I purposely removed `license`, and fixed it in the 2nd

![Screenshot 2023-09-23 at 09 59 41](https://github.com/node-red/flow-library/assets/55892693/75e08d6c-6437-4b23-9a2c-e3e462281fdb)

![Screenshot 2023-09-23 at 10 04 04](https://github.com/node-red/flow-library/assets/55892693/3f9c7914-726c-41e3-9586-0e30e796cc97)

What I did notice - `delete require.cache[require.resolve(...)]` appears after the `require`, not sure if that is key

at least that is what works for me.
https://github.com/marcus-j-davies/HAP-Router/blob/ba7f8d472fe28d74a011a6f6d9e390e3d0338392/core/server.js#L552

**Comment by sammachin** [comment]
Just released a couple of updates to my matter-bridge node this morning, first one 0.0.6 didn't have any changes to the scorecard compared to 0.0.5 (correct), then published 0.0.8 about 3 hours later where I'd added examples and the scorecard has been updated. 
So seems to be working for me at least
![Screen Shot 2023-10-09 at 1 10 01 PM](https://github.com/node-red/flow-library/assets/187645/c3252492-ab38-44de-9a30-f797fa48dde3)
![Screen Shot 2023-10-09 at 1 09 56 PM](https://github.com/node-red/flow-library/assets/187645/246bea91-00e2-49e7-b7a4-5022c1f450dc)
